llvm/lib/Target/AMDGPU/SIInstrInfo.cpp

```diff
@@ -4214,6 +4214,13 @@ bool SIInstrInfo::isSchedulingBoundary(const MachineInstr &MI,
   if (MI.getOpcode() == AMDGPU::SCHED_BARRIER && MI.getOperand(0).getImm() == 0)
     return true;
 
+  // The scheduler does not understand what kind of external events this
+  // instruction waits for, so cannot do a good job of scheduling it. Making it
+  // a boundary allows front ends to insert it at an appropriate place without
```

Comment:
Frontends could achieve the same thing with `sched_barrier(0)`, but having any kind of scheduling barrier would prevent independent non-sideeffect having instructions from being scheduled before the `s_wait_event` so it will not be as late as possible.

You could add a custom heuristic in tryCand, or there is probably some trick with the SchedModel if you make it occupy a resource for x cycles, but if what you have works well enough then LGTM.

Reply:
> Frontends could achieve the same thing with `sched_barrier(0)`

Good point. I might try that instead.

> You could add a custom heuristic in tryCand

Yeah, that would be my next approach, but I am reluctant because we have several different cut-n-pasted copies of `tryCandidate` which should all have the same tweak.

> but if what you have works well enough then LGTM.

Thanks.